src/cobald/daemon/runners/asyncio_runner.py

```diff
@@ -1,81 +1,69 @@
+        super().__init__(asyncio_loop)
         self._tasks = set()
+        self._failure_queue = asyncio.Queue()
 
-    def register_payload(self, payload):
-        super().register_payload(partial(raise_return, payload))
+    def register_payload(self, payload: Callable[[], Awaitable]):
+        self.asyncio_loop.call_soon_threadsafe(self._setup_payload, payload)
 
-    def run_payload(self, payload):
-        execution = AsyncExecution(payload)
-        super().register_payload(execution.coroutine)
-        return execution.wait()
+    def run_payload(self, payload: Callable[[], Coroutine]):
+        future = asyncio.run_coroutine_threadsafe(payload(), self.asyncio_loop)
+        return future.result()
 
-    def _run(self):
-        asyncio.set_event_loop(self.event_loop)
-        self.event_loop.run_until_complete(self._run_payloads())
+    def _setup_payload(self, payload: Callable[[], Awaitable]):
+        task = self.asyncio_loop.create_task(self._monitor_payload(payload))
+        self._tasks.add(task)
 
-    async def _run_payloads(self):
-        """Async component of _run"""
-        delay = 0.0
+    async def _monitor_payload(self, payload: Callable[[], Awaitable]):
         try:
-            while self.running.is_set():
-                await self._start_payloads()
-                await self._reap_payloads()
-                await asyncio.sleep(delay)
-                delay = min(delay + 0.1, 1.0)
-        except Exception:
-            await self._cancel_payloads()
+            result = await payload()
+        except asyncio.CancelledError:
             raise
+        except BaseException as e:
+            failure = e
+        else:
+            if result is None:
+                return
+            failure = OrphanedReturn(payload, result)
+        finally:
+            self._tasks.discard(asyncio_current_task())
+        await self._failure_queue.put(failure)
 
-    async def _start_payloads(self):
-        """Start all queued payloads"""
-        with self._lock:
-            for coroutine in self._payloads:
-                task = self.event_loop.create_task(coroutine())
-                self._tasks.add(task)
-            self._payloads.clear()
-        await asyncio.sleep(0)
-
-    async def _reap_payloads(self):
-        """Clean up all finished payloads"""
-        for task in self._tasks.copy():
-            if task.done():
-                self._tasks.remove(task)
-                if task.exception() is not None:
-                    raise task.exception()
-        await asyncio.sleep(0)
-
-    async def _cancel_payloads(self):
-        """Cancel all remaining payloads"""
-        for task in self._tasks:
-            task.cancel()
-            await asyncio.sleep(0)
-        for task in self._tasks:
-            while not task.done():
-                await asyncio.sleep(0.1)
-                task.cancel()
+    async def manage_payloads(self):
+        failure = await self._failure_queue.get()
```

Comment:
The queue currently is not really handled as a queue, so it is not possible to wait for several failures that might happen at the same time.
I see several options here, either using it as a future when only being interested in one failure (or ensuring, that none happened) or using it as a queue and checking if there are more than just one failure.

Reply:
This is somewhat tricky since "multiple failures" isn't really something that Python supports. [Exception groups are only upcoming in 3.11](https://docs.python.org/3.11/whatsnew/3.11.html#summary-release-highlights) and until there is no standard for this.

So I'm hesitant to roll our own "multiple failures" solution, since there will be heavy maintenance cost and tooling won't get any advantage from it (at most, we get super long traceback).  

I'll have to play around to see if using a future directly is simple enough.

Reply:
Turns out a `Future` is reasonably simple. It adds one check (to see if there is a result/failure already) but should still be much lighter than a queue.

I've changed both the asyncio and threading runners to use this mechanism.